File: Python/ExampleTests/record_filter_example_glue.py
from common import *
from production import IDForm, IDValue, RecordFilter, TemperatureConverter


class RecordFilterExampleGlue:
    DNC_STRING = '?DNC?'

    def __init__(self):
        self.record_filter = RecordFilter()
        self.computed_sum = 0

    def given_list_of_numbers(self, values: list):
        self.record_filter = RecordFilter()
        for value in values:
            typed = IDValueTyped.from_string_obj(value)
            self.record_filter.add(IDValue(IDForm(typed.id), typed.value))

    def when_filtered_by_id_with_value(self, values: list):
        if values and values[0]:
            self.computed_sum = self.record_filter.sum_by_label(IDForm(values[0][0]))

    def then_sum_is(self, values: list):
        if values and values[0]:
            assert int(values[0][0]) == self.computed_sum, 'Sum'

    def when_filtered_by(self, values: list):
        for value in values:
            typed = FilterValueTyped.from_string_obj(value)
            self.computed_sum = self.record_filter.sum_by_label(IDForm(typed.value))

    def then_result(self, values: list):
        for value in values:
            typed = ResultValueTyped.from_string_obj(value)
            assert typed.sum == self.computed_sum, 'Filtered sum'

    def when_element_added(self, values: list):
        for value in values:
            typed = IDValueTyped.from_string_obj(value)
            self.record_filter.add(IDValue(IDForm(typed.id), typed.value))

    def examples_Calculation_ConvertFToC(self, values: list):
        for value in values:
            typed = FandCTyped.from_string_obj(value)
            actual = TemperatureConverter.fahrenheit_to_celsius(typed.f)
            assert typed.c == actual, f'Convert {typed.f}F to C'

    def examples_DataType_IDForm(self, values: list):
        for value in values:
            vvt = ValidValuesTyped.from_string_obj(value)
            failed = False
            try:
                IDForm(vvt.value)
            except ValueError:
                failed = True
            assert vvt.is_valid == (not failed), f' Value {vvt.value}'

File: Swift/ExampleTests/RecordFilterExampleGlue.swift
import XCTest

public class RecordFilterExampleGlue {
    public init() {}

    private var recordFilter = RecordFilter()
    private var computedSum = 0

    public func givenListOfNumbers(_ values: [IDValueString]) {
        recordFilter = RecordFilter()
        for value in values {
            let typed = IDValueTyped(from: value)
            guard let id = try? IDForm(typed.iD) else {
                return XCTFail("bad ID \(typed.iD)")
            }
            recordFilter.add(IDValue(id: id, value: typed.value))
        }
    }

    public func whenFilteredByIdWithValue(_ values: [[String]]) {
        guard let first = values.first?.first else { return }
        guard let id = try? IDForm(first) else { return XCTFail("bad ID \(first)") }
        computedSum = recordFilter.sumByLabel(id)
    }

    public func thenSumIs(_ values: [[String]]) {
        guard let first = values.first?.first,
              let expected = Int(first.trimmingCharacters(in: .whitespaces)) else { return }
        XCTAssertEqual(expected, computedSum, "Sum")
    }

    public func whenFilteredBy(_ values: [FilterValueString]) {
        for value in values {
            let typed = FilterValueTyped(from: value)
            guard let id = try? IDForm(typed.value) else {
                return XCTFail("bad ID \(typed.value)")
            }
            computedSum = recordFilter.sumByLabel(id)
        }
    }

    public func thenResult(_ values: [ResultValueString]) {
        for value in values {
            let typed = ResultValueTyped(from: value)
            XCTAssertEqual(typed.sum, computedSum, "Filtered sum")
        }
    }

    public func whenElementAdded(_ values: [IDValueString]) {
        for value in values {
            let typed = IDValueTyped(from: value)
            guard let id = try? IDForm(typed.iD) else {
                return XCTFail("bad ID \(typed.iD)")
            }
            recordFilter.add(IDValue(id: id, value: typed.value))
        }
    }

    public func examplesCalculationConvertFToC(_ values: [FandCString]) {
        for value in values {
            let typed = FandCTyped(from: value)
            XCTAssertEqual(typed.c, fahrenheitToCelsius(typed.f),
                           "Convert \(typed.f)F to C")
        }
    }

    public func examplesDataTypeIDForm(_ values: [ValidValuesString]) {
        for value in values {
            let vvt = ValidValuesTyped(from: value)
            var failed = false
            do { _ = try IDForm(vvt.value) } catch { failed = true }
            XCTAssertEqual(vvt.isValid, !failed, " Value \(vvt.value)")
        }
    }
}
